src/box/tuple_constraint_fkey.c

```diff
@@ -0,0 +1,349 @@
+
+	const char *key = mp_data;
+
+	const char *unused;
+	if (key_validate_parts(key_def, key, part_count, false, &unused) != 0) {
+		field_foreign_key_failed(constr, field, "wrong key type");
+		return -1;
+	}
+	struct tuple *tuple = NULL;
+	if (index_get(index, key, part_count, &tuple) != 0) {
+		field_foreign_key_failed(constr, field, "index get failed");
+		return -1;
+	}
+	if (tuple == NULL) {
+		field_foreign_key_failed(constr, field,
+					 "foreign tuple was not found");
+		return -1;
+	}
+	return 0;
+}
+
+/**
+ * Set diag error ER_FOREIGN_KEY_INTEGRITY with given message.
+ */
+static void
+foreign_key_integrity_failed(const struct tuple_constraint *constr,
+			     const char *message)
+{
+	struct error *err = diag_set(ClientError, ER_FOREIGN_KEY_INTEGRITY,
+				     constr->def.name, message);
+	error_set_str(err, "name", constr->def.name);
+}
+
+int
+tuple_constraint_fkey_check_delete(const struct tuple_constraint *constr,
+				   struct tuple *deleted_tuple,
+				   struct tuple *replaced_with_tuple)
+{
+	assert(deleted_tuple != NULL);
+	int32_t foreign_field_no = constr->fkey->data->foreign_field_no;
+	if (foreign_field_no < 0) {
+		foreign_key_integrity_failed(constr,
+					     "wrong foreign field name");
+		return -1;
+	}
+	if (replaced_with_tuple != NULL) {
+		/*
+		 * We have to check whether the tuple is replaced by equal in
+		 * terms of that foreign key constraint - if it is so then the
+		 * integrity cannot be broken and no more checks are needed.
+		 * Note that if there are no foreign index - the foreign key
+		 * constraint is already is in uncheckable state and perhaps
+		 * we may not check replaces tuple too.
+		 */
+		struct space *fspace = constr->space_cache_holder.space;
+		if (constr->fkey->foreign_index == 0) {
+			/*
+			 * The index in the foreign space is primary.
+			 * replaced_with_tuple is equal to deleted_tuple in
+			 * terms of the primary index.
+			 */
+			return 0;
+		} else if (constr->fkey->foreign_index > 0) {
+			/*
+			 * Just compare using foreign index's def.
+			 */
+			int32_t findex_no = constr->fkey->foreign_index;
+			struct index *findex = fspace->index[findex_no];
+			struct key_def *fkey_def = findex->def->key_def;
+			if (tuple_compare(deleted_tuple, HINT_NONE,
+					  replaced_with_tuple, HINT_NONE,
+					  fkey_def) == 0)
+				return 0;
+		}
+	}
+
+	if (constr->fkey->local_index < 0) {
+		foreign_key_integrity_failed(constr, "index was not found");
+		return -1;
+	}
+
+	struct index *index = constr->space->index[constr->fkey->local_index];
+	struct key_def *key_def = index->def->key_def;
+	uint32_t part_count = constr->fkey->field_count;
+	assert(constr->fkey->field_count == key_def->part_count);
+
+	const char *key = tuple_field(deleted_tuple, foreign_field_no);
+	if (key == NULL || mp_typeof(*key) == MP_NIL) {
+		/* No field - nobody can be bound to it.*/
+		return 0;
+	}
+
+	const char *unused;
+	if (key_validate_parts(key_def, key, part_count, false, &unused) != 0) {
```

Comment:
Why do you need to validate the key? Shouldn't the foreign and local index be compatible in the first place?

Reply:
The don't need to be compatible.